EDK2 on RISC-V Sophgo SG2042 platform

caiyuqing379 opened the pull request:
EDK2 on RISC-V Sophgo SG2042 platform

Description:
Deploy EDK2 to run on 64-core CPU under RISC-V architecture, and successfully boot to OS.

Implementation Details:
1. The startup process is roughly: ZSBL+FSBL+OpenSBI+EDK2+GRUB+OS
2. The running platform is the SG2042 platform of the RISC-V architecture of the 64-core C920.
3. Adopted the scheme of separating OpenSBI and EDK2
Testing:
The project completed the test of the EVB version of Sophgo SG2042
video:
https://drive.google.com/file/d/15iT-qkSYf6jiZsZV3bsGRJN0MAh2USDK/view

I'm looking forward to your feedback and suggestions, and I hope this feature will be a beneficial addition to the project.

Thank you for your review!

Signature: caiyuqing379 - <202235273@mail.sdu.edu.cn>

vlsunil commented:
This is great!. Glad to see you adopted payload option. Let me have a closer look.

vlsunil reviewed (commented):
Do you have any changes in SEC module compared to that of RiscVVirt? If not, is it possible to reference RIscVVirt's SEC directly in your DSC/DEC files instead of copying? When I had added the SEC in RiscVVirt, the primary reason was it had dependency on FDT library. Now there is a new FDT library in MdePkg which will allow me to create SEC under UefiCpuPkg. So, in future we may move some of these generic things to UefiCpuPkg or MdePkg in which case you don't need to have copy anyway. So, referencing RiscVVirt may be better in the interim.

vlsunil commented on Platform/Sophgo/SG2042Pkg/Library/PeiCoreInfoHobLib/PeiCoreInfoHobLib.inf:
```diff
@@ -0,0 +1,44 @@
+## @file
+#  Library instance to create core information HOB
```
I am not sure whether you really this and RiscVSmbiosDxe. type 44 specific to RISC-V is not ratified / frozen table (though referenced from smbios spec) and the current info what it provides already provided by other means now. So, see if you can avoid these modules by using standard SMBIOS module.

caiyuqing379 replied:
Thank you for your suggestion! We will try to avoid these modules by using standard SMBIOS modules.

vlsunil reviewed (commented):
Added some high level comments. Once you decide on them, you can send patches to edk2-devel.

caiyuqing379 commented:
> Do you have any changes in SEC module compared to that of RiscVVirt? If not, is it possible to reference RIscVVirt's SEC directly in your DSC/DEC files instead of copying? When I had added the SEC in RiscVVirt, the primary reason was it had dependency on FDT library. Now there is a new FDT library in MdePkg which will allow me to create SEC under UefiCpuPkg. So, in future we may move some of these generic things to UefiCpuPkg or MdePkg in which case you don't need to have copy anyway. So, referencing RiscVVirt may be better in the interim.

Hello! Added the BuildCoreInformationHob function in SecMain. c to build the core information hob. Two changes have been made in Memory. c. The memory space size of SG2042 EVB is determined by the number and size of DDRs inserted on the board. There is a problem with initializing the system ram space of each memory node separately using InitializeRamRegions, so InitializeRamRegions is only called once for total system ram initialization. At the same time, add the AddRuntimeServicesMemoryMap function to mark runtime memory ranges in the EFI memory map.

We will add some comments in the code later to explain these issues!

caiyuqing379 commented on Platform/Sophgo/SG2042Pkg/Universal/Dxe/FdtDxe/FdtDxe.c:
```diff
@@ -0,0 +1,150 @@
+  mBootHartId = FirmwareContext->BootHartId;
+  DEBUG ((DEBUG_INFO, " %a: mBootHartId = 0x%x.\n", __func__, mBootHartId));
+
+  Status = FixDtb (DataInHob, mBootHartId);
```
Sorry, there is still a method for getting boot hard ID from FDT in the Linux kernel, so this dxe driver has been kept. We will remove it later.

caiyuqing379 commented on Platform/Sophgo/SG2042Pkg/Universal/Dxe/RamFvbServicesRuntimeDxe/FvbServicesRuntimeDxe.inf:
```diff
@@ -0,0 +1,84 @@
+## @file
+#  Component description file for RAM Flash Fimware Volume Block DXE driver
```
Thank you for your feedback! The FvbServicesRuntimeDxe is required by VariableRuntimeDxe. And some modifications have been made to FwBlockService. c in the original module. The MarkMemoryRangeForRuntimeAccess function has been removed because it conflicts with memory initialization in the Sec module. We will try to see if there is a better solution.

caiyuqing379 replied:
Thank you for your feedback! We will no longer be adding variable memory to the SEC stage.

vlsunil replied:
That's great!. Thank you.

vlsunil commented on Platform/Sophgo/SG2042Pkg/Library/RiscVMmuLib/BaseRiscVMmuLib.inf:
```diff
@@ -0,0 +1,28 @@
+## @file
+#  RISC-V MMU library.
```
I am still not sure why do you need to copy this library.... Is there something which deviates from standard RISC-V specs? Again, just trying to avoid duplication.

vlsunil commented on Platform/Sophgo/SG2042Pkg/About_Sophgo_platform.md:
```diff
@@ -0,0 +1,40 @@
+## Sophgo SG2042 Platform
+The development work for this project was carried out around the Sophgo SG2042 EVB development board. 
+The deployment of EDK2 was completed on the EVB test version. Below is the external appearance display of the EVB version.  
+![image](https://github.com/AII-SDU/EDK2_SG2042/blob/separate-v2/edk2-platforms/Platform/Sophgo/SG2042Pkg/Documents/Media/Sophgo_SG2042_EVB.png)  
```
Please check these links. I am unable to access.

caiyuqing379 replied:
Sorry, this is a link to a private repository, we will change the link to a public repository.

vlsunil replied:
No worries. May be you can provide relative link to the image in the same directory of the branch so that it always works for any branch you raise the PR.

vlsunil reviewed (commented):
Left few comments. Primarily I am looking to avoid code duplication as much as possible so that it is easier to maintain. Thanks a lot for this work. This will be great for the community.

caiyuqing379 commented on Platform/Sophgo/SG2042Pkg/Universal/Dxe/FdtDxe/FdtDxe.c:
```diff
@@ -0,0 +1,89 @@
+  DataInHob = (VOID *)*((UINTN *)GET_GUID_HOB_DATA (GuidHob));
+  DataSize  = GET_GUID_HOB_DATA_SIZE (GuidHob);
+
+  Status = gBS->InstallConfigurationTable (&gFdtTableGuid, DataInHob);
```
Thank you for your suggestion, to avoid duplication we will remove FdtDxe. But to use FdtClientDxe you need to install gEdkiiPlatformHasDeviceTreeGuid, so a PlatformHasDtDxe is required.

vlsunil replied:
Yes, Is that an issue? OTOH, it would be better when you want to support ACPI.

caiyuqing379 commented on Platform/Sophgo/SG2042Pkg/Sec/SecEntry.S:
```diff
@@ -0,0 +1,21 @@
+#include "SecMain.h"
+
+.text
+.align 3
```
Sorry, we are not currently using clang to build edk2, and will track down the latest code.

caiyuqing379 commented on Platform/Sophgo/SG2042Pkg/Sec/Memory.c:
```diff
@@ -0,0 +1,400 @@
+  and size of DDRs inserted on the board. There is an error with initializing 
+  the system ram space of each memory node separately using InitializeRamRegions, 
+  so InitializeRamRegions is only called once for total system ram initialization.
+
```
Sure! I'll send you this error when it's sorted out.
